src/mixins/java/org/spongepowered/common/mixin/core/recipebook/ServerPlaceRecipeMixin.java

```diff
@@ -0,0 +1,126 @@
+import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
+import net.minecraft.core.Holder;
+import net.minecraft.recipebook.ServerPlaceRecipe;
+import net.minecraft.world.entity.player.Inventory;
+import net.minecraft.world.entity.player.StackedItemContents;
+import net.minecraft.world.inventory.Slot;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.RecipeHolder;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.common.bridge.world.item.crafting.PlacementInfoBridge;
+import org.spongepowered.common.item.recipe.book.RecipeBookUtil;
+import org.spongepowered.common.item.recipe.book.SpongeStackedItemContents;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Makes recipe placing to work with sponge custom ingredients.
+ * Fallbacks to the original logic if recipe does not have any custom ingredient.
+ */
+@Mixin(ServerPlaceRecipe.class)
+public abstract class ServerPlaceRecipeMixin {
+
+    @Shadow @Final private Inventory inventory;
+
+    private @Nullable List<ItemStack> impl$stackList;
+
+    @WrapOperation(
+        method = "placeRecipe(Lnet/minecraft/recipebook/ServerPlaceRecipe$CraftingMenuAccess;IILjava/util/List;Ljava/util/List;Lnet/minecraft/world/entity/player/Inventory;Lnet/minecraft/world/item/crafting/RecipeHolder;ZZ)Lnet/minecraft/world/inventory/RecipeBookMenu$PostPlaceAction;",
+        at = @At(
+            value = "NEW",
+            target = "()Lnet/minecraft/world/entity/player/StackedItemContents;"
+        )
+    )
+    private static StackedItemContents impl$useCustomStackedItemContents(
+        final Operation<StackedItemContents> original,
+        final ServerPlaceRecipe.CraftingMenuAccess<?> menu,
+        final int gridWidth, final int gridHeight,
+        final List<Slot> inputGridSlots, final List<Slot> slotsToClear,
+        final Inventory inventory, final RecipeHolder<?> recipe,
+        final boolean useMaxItems, final boolean isCreative
+    ) {
+        return ((PlacementInfoBridge) recipe.value().placementInfo()).bridge$hasCustomIngredients()
+            ? new SpongeStackedItemContents()
+            : original.call();
+    }
+
+    @WrapMethod(method = "placeRecipe(Lnet/minecraft/world/item/crafting/RecipeHolder;Lnet/minecraft/world/entity/player/StackedItemContents;)V")
+    private void impl$handleApplicableStacks(
+        final RecipeHolder<?> recipe, final StackedItemContents contents, final Operation<Void> original
+    ) {
+        if (contents instanceof final SpongeStackedItemContents spongeContents) {
+            this.impl$stackList = new ArrayList<>();
```

Comment:
```suggestion
            this.impl$stackList = new ArrayDequeue<>();
```
Removing the first element from a list is O(n)